x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosRealm.java

```diff
@@ -215,18 +215,17 @@ private void resolveUser(final String userPrincipalName, final String outToken,
             final User user = (userPrincipalNameToUserCache != null) ? userPrincipalNameToUserCache.get(username) : null;
             if (user != null) {
                 listener.onResponse(AuthenticationResult.success(user));
+            } else if (userAndRealmName.length > 1) {
+                final String realmName = userAndRealmName[1];
+                buildUser(username, Map.of(KRB_METADATA_REALM_NAME_KEY, realmName, KRB_METADATA_UPN_KEY, userPrincipalName), listener);
             } else {
-                final String realmName = (userAndRealmName.length > 1) ? userAndRealmName[1] : null;
-                final Map<String, Object> metadata = new HashMap<>();
-                metadata.put(KRB_METADATA_REALM_NAME_KEY, realmName);
-                metadata.put(KRB_METADATA_UPN_KEY, userPrincipalName);
-                buildUser(username, metadata, listener);
+                buildUser(username, Map.of(KRB_METADATA_UPN_KEY, userPrincipalName), listener);
             }
         }
     }
 
     private void buildUser(final String username, final Map<String, Object> metadata, final ActionListener<AuthenticationResult> listener) {
-        final UserRoleMapper.UserData userData = new UserRoleMapper.UserData(username, null, Collections.emptySet(), metadata, this.config);
+        final UserRoleMapper.UserData userData = new UserRoleMapper.UserData(username, null, Set.of(), metadata, this.config);
```

Comment:
Is there a reason to change this? I don't feel strongly, but I do prefer `emptySet` as being more readable in a standalone context.

Reply:
Yes, there is a reason. I try to be consistent and return an `AbstractImmutableCollection` for every return path in the function.

Background:

The goal of the PR is to use the new API for immutable collections. They have two advantages: improved API for building small  "unmodifiable" collections (`ImmutableCollections`) and preventing wrapping of an `unmodifiable` collection by another `unmodifiable` collection (see for example `java.util.Set#copyOf`). The "drawback" is that `null` elements are not permitted.

What I want to achieve in this PR is to construct `ImmutableCollections` instances as soon as possible (using `Set.of`, `Set.copyOf` and `Collectors.toUnmodifiableSet()` - equivalent for lists) - at the interface with another library - and then pass those instances up the call chain without further "unmodifiable" wrappings or checks for `null` elements during iterations. At the lowest level, the interface with another library, I looked at their code to check if they have `null`-checks in place; if they did I used the `Set.of` or `Set.copyOf` functions, because if they catch a `null` element and throw then we have catched a bug in the library; otherwise, if there are no `null`-checks in the library, I simply filtered out the `null` elements (they were not relevant) using a `Stream` and then used the `Collectors.toUnmodifiableSet` collector.

Unfortunately, the type system is not helping us in guaranteeing that the parameter is such an "unmodifiable" collection, so `Set.copyOf` or `List.copyOf` are to be expected in the code, between "abstraction layers", but they will not create new instances if indeed the Collection was `AbstractImmutableCollection` (so, if it was created with `Set.of` instead of `Collections.emptySet()`), so they are cheap.